src/levanter/optim/config.py

```diff
@@ -139,27 +143,39 @@ def mask_fn(model):
             return mask_fn
 
     def lr_scheduler(self, num_train_steps):
-        warmup_steps = self._convert_warmup(num_train_steps)
-        cooldown_steps = _convert_ratio_or_steps(self.cooldown, num_train_steps)
-        if self.haps is None:
-            self.haps = []
-        self.haps.insert(0, warmup_steps)
-        self.haps.append(num_train_steps - cooldown_steps)
+        if self.cooldown is not None:
+            warnings.warn("cooldown is deprecated. Just use the normal schedule.", DeprecationWarning)
+            cooldown_steps = _convert_ratio_or_steps(self.cooldown, num_train_steps)
+        else:
+            cooldown_steps = 0
+
+        total_main_steps = num_train_steps - cooldown_steps
+        cooldown_points = self._get_cycle_minima(total_main_steps)
+
+        cooldown_points.insert(0, 0)
+        cooldown_points.append(num_train_steps)
 
         min_lr = self.learning_rate * self.min_lr_ratio
 
         schedules = []
         boundaries = []
 
-        if warmup_steps != 0:
-            warmup = optax.linear_schedule(0.0, self.learning_rate, warmup_steps)
-            schedules.append(warmup)
-            boundaries.append(warmup_steps)
+        previous_end = 0.0
 
-        for start, end in zip(self.haps[:-1], self.haps[1:]):
+        for cycle, (start, end) in enumerate(zip(cooldown_points[:-1], cooldown_points[1:])):
             cycle_steps = end - start
+            if cycle == 0:  # warmup
+                warmup_steps = _convert_ratio_or_steps(self.warmup, cycle_steps)
```

Comment:
should this be `total_steps` or `cycle_steps`?

Reply:
I think it's easier to handle `cycle_steps` cuz it ensures `warmup_steps <= cycle_steps`. But I also think `self.warmup` is a global schedule thing not a cycle schedule thing.

Reply:
makes sense.